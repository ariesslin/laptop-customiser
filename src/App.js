import React, {useCallback, useEffect, useState} from "react";
import Variant from "./Variant";
import {getDefaultPrice} from "./service";

const App = () => {
  const [loading, setLoading] = useState(false);
  const [processorList, setProcessorList] = useState([]);
  const [error, setError] = useState(null);
  const [priceLoading, setPriceLoading] = useState(false);
  const [price, setPrice] = useState(0);
  const [priceError, setPriceError] = useState(null);
  const [selectedVariantSerialNo, setSelectedVariantSerialNo] = useState(
      {
    proe: ''
  })

  useEffect(() => {
    const getMacComponents = () => {
      return fetch(`http://localhost:3004/components`)
        .then((response) => response.json())
        .then((data) => {
          const processorComponents = data.filter(component => component.type === "proe");
          setSelectedVariantSerialNo({
            proe: processorComponents.find(component => component.default).serialNo
          });
          setProcessorList(processorComponents);
          setLoading(false);
        })
        .catch(() => {
          setError(new Error("could not fetch the customisable components"));
          setLoading(false);
        });
    };
    setLoading(true);
    getMacComponents();
    setPriceLoading(true);
    getDefaultPrice()
      .then((data) => {
        setPrice(data.value);
      })
      .catch((error) => {
        setPriceError(error);
      })
      .finally(() => {
        setPriceLoading(false);
      });
  }, []);

  const setSelectedVariant = useCallback( (variantSerialNo) => {
    setSelectedVariantSerialNo({
      proe: variantSerialNo
    })
  }, []);

  const getSelectedVariant = () => {
    return processorList.find(processor => processor.serialNo === selectedVariantSerialNo.proe);
  }

  const getAddOnPrice = () => {
    return getSelectedVariant()?.addOnPrice ?? 0;
  };

  return (
    <>
      <header>
        <div className="header__content">
          <a
            className="header__link"
            href="https://www.apple.com/in/macbook-pro"
          >
            <strong>MacBook Pro</strong>
          </a>
        </div>
      </header>
      <main>
        <div className="main__container">
          <div className="main__content">
            <section>
              <img
                className="macbook-img"
                alt="macbook pro"
                src="https://store.storeimages.cdn-apple.com/4668/as-images.apple.com/is/mbp16touch-space-select-201911?wid=1808&hei=1686&fmt=jpeg&qlt=80&.v=1572825197207"
              />
            </section>
            <section className="configuration">
              {loading ? (
                <h1>loading...</h1>
              ) : error ? (
                <h1>Something went wrong. Please try again later</h1>
              ) : (
                <>
                  <h1 className="mt-0">
                    Customise your 16‑inch MacBook Pro - Space Grey
                  </h1>
                  <ul className="summary-list">
                    <li data-testid="processor_desc">{getSelectedVariant()?.variant}</li>
                    <li>16-inch Retina display with True Tone</li>
                    <li>Four Thunderbolt 3 ports</li>
                    <li>Touch Bar and Touch ID</li>
                    <li>Backlit Magic Keyboard - US English</li>
                  </ul>
                  {!!processorList.length && (
                    <Variant
                        name="Processor"
                        processorList={processorList}
                        setSelectedVariant={setSelectedVariant}
                        selectedVariantSerialNo={selectedVariantSerialNo.proe}
                    />
                  )}
                </>
              )}
            </section>
          </div>
        </div>
        <div className="price container">
          <div className="price__content">
            <h1 className="price__value">
              Total:
              {priceLoading ? (
                <>{` loading...`}</>
              ) : priceError ? (
                "Something went wrong. Please try again later"
              ) : (
                <span data-testid="total-price">{` ₹${
                  price + getAddOnPrice()
                }`}</span>
              )}
            </h1>
          </div>
        </div>
      </main>
    </>
  );
};

export default App;
