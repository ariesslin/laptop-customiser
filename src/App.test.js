import { fireEvent, render, waitFor } from "@testing-library/react";
import App from "./App";
import data from "../server/db.json";

describe("Laptop customiser", () => {
  beforeEach(() => {
    jest
      .spyOn(global, "fetch")
      .mockResolvedValueOnce({
        json: () => Promise.resolve(data.components),
      })
      .mockResolvedValueOnce({
        json: () => Promise.resolve(data.price),
      });
  });

  it("should show the correct price when processor a is selected", async () => {
    const { getByTestId } = render(<App />);
    await waitFor(() => {
      fireEvent.click(getByTestId(`Processor_a`));
    })
    expect(getByTestId("total-price")).toHaveTextContent(`₹239900`);
  });

  it("should show the correct price when processor b is selected", async () => {
    const { getByTestId } = render(<App />);
    await waitFor(() => {
      fireEvent.click(getByTestId(`Processor_b`));
    });
    expect(getByTestId("total-price")).toHaveTextContent(`₹259900`);
  });

  it("should show the list of processors", async () => {
    const { getByTestId } = render(<App />);
    await waitFor( () => {
      expect(getByTestId(`Processor_a`)).toBeVisible();
      expect(getByTestId(`Processor_b`)).toBeVisible();
    }) ;
  });

  it("should get the first processor selected when page is loaded", async () => {
    const { getByTestId } = render(<App />);
    await waitFor(() => {
      expect(getByTestId("Processor_a")).toHaveClass("variant--selected");
    });
  });

  it("should get the processor b selected when processor b is clicked", async () => {
    const { getByTestId } = render(<App />);
    await waitFor(() => {
      fireEvent.click(getByTestId(`Processor_b`));
      expect(getByTestId("Processor_b")).toHaveClass("variant--selected");
    });
  });

  it("should get the first processor description when page is loaded", async () => {
    const { getByTestId } = render(<App />);
    await waitFor(() => {
      expect(getByTestId("processor_desc")).toHaveTextContent("2.3GHz 8-core 9th-generation Intel Core processor, Turbo Boost up to 4.8GHz");
    });
  });

  it("should get the selected processor b description when processor b is clicked", async () => {
    const { getByTestId } = render(<App />);
    await waitFor(() => {
      fireEvent.click(getByTestId(`Processor_b`));
        expect(getByTestId("processor_desc")).toHaveTextContent("2.4GHz 8-core 9th-generation Intel Core processor, Turbo Boost up to 5.0GHz");
    });
  });

  it("should match snapshot after processor_a is clicked", async () => {
    const { container, getByTestId } = render(<App />);
    await waitFor(() => {
      fireEvent.click(getByTestId(`Processor_a`));
    });
    expect(container).toMatchSnapshot();
  });
});
