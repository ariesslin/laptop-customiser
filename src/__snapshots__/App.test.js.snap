// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Laptop customiser should match snapshot 1`] = `
<div>
  <header>
    <div
      class="header__content"
    >
      <a
        class="header__link"
        href="https://www.apple.com/in/macbook-pro"
      >
        <strong>
          MacBook Pro
        </strong>
      </a>
    </div>
  </header>
  <main>
    <div
      class="main__container"
    >
      <div
        class="main__content"
      >
        <section>
          <img
            alt="macbook pro"
            class="macbook-img"
            src="https://store.storeimages.cdn-apple.com/4668/as-images.apple.com/is/mbp16touch-space-select-201911?wid=1808&hei=1686&fmt=jpeg&qlt=80&.v=1572825197207"
          />
        </section>
        <section
          class="configuration"
        >
          <h1>
            loading...
          </h1>
        </section>
      </div>
    </div>
    <div
      class="price container"
    >
      <div
        class="price__content"
      >
        <h1
          class="price__value"
        >
          Total:
           loading...
        </h1>
      </div>
    </div>
  </main>
</div>
`;

exports[`Laptop customiser should match snapshot after processor_a is clicked 1`] = `
<div>
  <header>
    <div
      class="header__content"
    >
      <a
        class="header__link"
        href="https://www.apple.com/in/macbook-pro"
      >
        <strong>
          MacBook Pro
        </strong>
      </a>
    </div>
  </header>
  <main>
    <div
      class="main__container"
    >
      <div
        class="main__content"
      >
        <section>
          <img
            alt="macbook pro"
            class="macbook-img"
            src="https://store.storeimages.cdn-apple.com/4668/as-images.apple.com/is/mbp16touch-space-select-201911?wid=1808&hei=1686&fmt=jpeg&qlt=80&.v=1572825197207"
          />
        </section>
        <section
          class="configuration"
        >
          <h1
            class="mt-0"
          >
            Customise your 16‑inch MacBook Pro - Space Grey
          </h1>
          <ul
            class="summary-list"
          >
            <li
              data-testid="processor_desc"
            >
              2.3GHz 8-core 9th-generation Intel Core processor, Turbo Boost up to 4.8GHz
            </li>
            <li>
              16-inch Retina display with True Tone
            </li>
            <li>
              Four Thunderbolt 3 ports
            </li>
            <li>
              Touch Bar and Touch ID
            </li>
            <li>
              Backlit Magic Keyboard - US English
            </li>
          </ul>
          <div
            class="component"
          >
            <h3
              class="component__name"
            >
              Processor
            </h3>
            <ul>
              <li
                class="variant variant--selected"
                data-testid="Processor_a"
              >
                <p
                  class="variant__name"
                >
                  <strong>
                    2.3GHz 8-core 9th-generation Intel Core processor, Turbo Boost up to 4.8GHz
                  </strong>
                </p>
              </li>
              <li
                class="variant "
                data-testid="Processor_b"
              >
                <p
                  class="variant__name"
                >
                  <strong>
                    2.4GHz 8-core 9th-generation Intel Core processor, Turbo Boost up to 5.0GHz
                  </strong>
                </p>
                <p>
                  + ₹
                  20000
                </p>
              </li>
            </ul>
          </div>
        </section>
      </div>
    </div>
    <div
      class="price container"
    >
      <div
        class="price__content"
      >
        <h1
          class="price__value"
        >
          Total:
          <span
            data-testid="total-price"
          >
             ₹239900
          </span>
        </h1>
      </div>
    </div>
  </main>
</div>
`;
